java/server/src/org/openqa/selenium/grid/distributor/local/LocalDistributor.java

```diff
@@ -88,33 +106,122 @@
+      NewSessionQueuer sessionRequests,
+      Secret registrationSecret,
+      Duration sessionRequestTimeout) {
     super(tracer, clientFactory, new DefaultSlotSelector(), sessions, registrationSecret);
+
     this.tracer = Require.nonNull("Tracer", tracer);
     this.bus = Require.nonNull("Event bus", bus);
     this.clientFactory = Require.nonNull("HTTP client factory", clientFactory);
     this.sessions = Require.nonNull("Session map", sessions);
     this.model = new GridModel(bus, registrationSecret);
     this.nodes = new HashMap<>();
+    this.sessionRequests = Require.nonNull("New Session Request Queue", sessionRequests);
+    this.sessionRequestTimeout = sessionRequestTimeout;
 
     bus.addListener(NodeStatusEvent.listener(status -> register(registrationSecret, status)));
     bus.addListener(NodeStatusEvent.listener(status -> model.refresh(registrationSecret, status)));
     bus.addListener(NodeDrainComplete.listener(this::remove));
+
+    bus.addListener(NewSessionRequestEvent.listener(reqId -> {
+      Optional<HttpRequest> sessionRequest = this.sessionRequests.remove();
+      // Check if polling the queue did not return null
+      if (sessionRequest.isPresent()) {
+        handleNewSessionRequest(sessionRequest.get(), reqId);
+      } else {
+        fireSessionRejectedEvent("Unable to poll request from the new session request queue.",
+                                 reqId);
+      }
+    }));
+  }
+
+  private void handleNewSessionRequest(HttpRequest sessionRequest, RequestId reqId) {
+
+    Span span = newSpanAsChildOf(tracer, sessionRequest, "distributor.poll_queue");
+    Map<String, EventAttributeValue> attributeMap = new HashMap<>();
+    attributeMap.put(AttributeKey.LOGGER_CLASS.getKey(),
+                     EventAttribute.setValue(getClass().getName()));
+    span.setAttribute(AttributeKey.REQUEST_ID.getKey(), reqId.toString());
+    attributeMap.put(AttributeKey.REQUEST_ID.getKey(), EventAttribute.setValue(reqId.toString()));
+
+    attributeMap.put("request", EventAttribute.setValue(sessionRequest.toString()));
+    try {
+      if (hasRequestTimedOut(sessionRequest)) {
+        span.addEvent("New session request timed out", attributeMap);
+        fireSessionRejectedEvent("New session request timed out", reqId);
+      } else {
+        try {
+          CreateSessionResponse sessionResponse = newSession(sessionRequest);
+          NewSessionResponse newSessionResponse =
+              new NewSessionResponse(reqId, sessionResponse.getSession(),
+                                     sessionResponse.getDownstreamEncodedResponse());
+
+          bus.fire(new NewSessionResponseEvent(newSessionResponse));
+        } catch (SessionNotCreatedException e) {
+          // If error is due to the slots being busy, adding to front of queue else reject the request
+          // If that fails, then just reject the request.
+          if (e.getMessage().startsWith("Unable to find provider for session") ||
```

Comment:
At this point, we may want `newSession` to return something more meaningful than what it's doing right now. If we had proper functional primitives, an `Either` might be useful, since we effectively want the response (if there is one) or a meaningful cause (if there isn't a response)

We then want to be able to respond differently based on the cause.

Reply:
Alright, thank you for the direction. That would definitely be a nice and clean way to do it. Will add a way to use an `Either` for the `newSession`.